man/man8/zpoolconcepts.8

```diff
@@ -93,6 +93,56 @@ compromised.
 The minimum number of devices in a raidz group is one more than the number of
 parity disks.
 The recommended number is between 3 and 9 to help increase performance.
+.It Sy draid , draid1 , draid2 , draid3
+A distributed version of raidz with integrated hot spares that allows for
+faster resilvering.
+A draid vdev is constructed from multiple internal raidz groups, each with D
+data devices and P parity devices, which are distributed over the children.
+This allows all of the children to participate when resilvering.
+As a general rule the more devices in a draid the less time required to
+resilver to a distributed spare.
+Furthermore, a narrower stripe width (D+P) will improve resilver performance
+at the expense of total usable capacity.
```

Comment:
I think we should talk about the minimum allocation unit for DRAID, and its relation to `D`, since this has a big effect on IOPS and capacity.  My understanding is that DRAID allocates in units of `D << ashift`.  For example, with the default `D=8` and 4K disks, it allocates in units of 32K.

The impact on IOPS is that for any read, `D` disks would be accessed.  However, since "groups" may overlap with each other, you can't do `D/<total data disks>` concurrent reads, since some will be from the same disk.  Maybe for simplicity we should assume that the entire DRAID group can do only the IOPS of a single drive?  It would be interesting to quantify this (either theoretically or experimentally).  For example, do we get `floor(D/<data disks>) * <single-drive-IOPS>`?  I would guess so, if we have enough outstanding random reads.

For capacity, the most obvious impacts would be with zvols, which would experience 4x space inflation with the default volblocksize=8k.  If using compression with the default recordsize=128K, the compression ratio would be reduced because each block is only 4 allocation units uncompressed, so if the compressed sizes are evenly distributed (mod 32K), we'll "waste" on average 16K per 128K (12.5%).

By adjusting `D`, these effects can be changed (larger `D` -> less IOPS, worse compression, slower resilver, but also less space used for parity).

It would be wonderful if we could automatically change the default volblocksize such that we will waste less than 50% of the space due to allocation round-up.  Similarly we could warn when manually setting volblocksize or recordsize to values that would waste half or more of the space.

I think we should also update zio_write_compress() to round up the compressed size to be a multiple of the smallest-allocation-unit device (in the normal class), for the same reason that we round up to a multiple of the smallest-ashift.

Reply:
I've updated this section of the man page to cover the capacity / IOPs considerations as they apply to D at a reasonably high level.  I wanted to try and avoid getting to specific in order to keep the dRAID section small enough to be easily read and understood.

> It would be wonderful if we could automatically change the default volblocksize such that we will waste less than 50% of the space due to allocation round-up. 

This was done in a previous change.  The default volblocksize is now set such that we always waste less than 50% of the space due to allocation round up.  With the default D=8 and 4k sectors disks the default volblocksize is 32k which ignoring compression is 100% used.

> I think we should also update zio_write_compress()

I still need to look in to this.